File: 2130. Maximum Twin Sum of a Linked List
/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        ListNode* dummy = head;
        ListNode* fast = head;
        ListNode* slow = head;
        int halfn = 0, res = 0;
        while (dummy != NULL && dummy -> next != NULL) {
            dummy = dummy -> next -> next;
            fast = fast -> next;
            halfn++;
        }
        vector<int> v(halfn, 0);
        for (int i = 0; i < halfn; i++) {
            v[halfn - i - 1] += fast -> val;
            v[i] += slow -> val;
            fast = fast -> next;
            slow = slow -> next;
            res = max(res, max(v[halfn - i - 1], v[i]));
        }
        return res;
    }
};

File: 2130. Maximum Twin Sum of a Linked List
/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        dummy, fast, slow = head, head, head
        halfn, res = 0, 0
        while dummy and dummy.next:
            dummy, fast = dummy.next.next, fast.next
            halfn += 1
        v = [0 for i in range(halfn)]
        for i in range(halfn):
            v[-i - 1] += fast.val
            v[i] += slow.val
            fast, slow = fast.next, slow.next
            res = max(res, v[-i - 1], v[i])
        return res